scenario-discovery (#4)
josephguillaume/uncertainfutures: _theory/PRIM.md

layout: concept
title: PRIM
permalink: /theory/PRIM/

> excerpt: PRIM (Patient Rule Induction Method) is a "bump-hunting" algorithm that identifies "boxes" (i.e., scenarios) in the uncertain input variable space within which the value of an output variable is optimised (i.e., minimised or maximised) relative to its average value over the while input space [4].

Reviewer: This is a bit unwieldy for the excerpt.
Boxes are not scenarios, and the method of identifying those boxes isn't crucial in this case (it doesn't differentiate the method from others)

purpose: PRIM can be used for <a href="../../theory/scenario-discovery/">scenario discovery</a> where the scenarios identified will be represented as bounds on parameter values within which scenarios meet some pre-defined criteria [1].

> requirements: <ol><li>Dataset generation from a simulation model</li><li>users' potential selection of candidate boxes for 'pasting'</li><li>users' selection of a box among candidate boxes for the 'covering' stage [1]</li></ol>

Reviewer: There isn't actually a heading for requirements, so introductory text is needed.
The first point could be written more simply.
Points 3 and 4 seem like steps in how to use PRIM, and need to have pasting and covering defined to make sense

requirements: <ol><li>Dataset generation from a simulation model</li><li>users' potential selection of candidate boxes for 'pasting'</li><li>users' selection of a box among candidate boxes for the 'covering' stage [1]</li></ol>
decomposition: >
  <p>Scenarios can be randomly or systematically sampled but need to cover the parameter space well to obtain sensible parameter bounds</p>
  <p>Scenarios are identified as interesting based on pre-defined criteria, e.g. scenarios where the selected decision alternative fails to meet required performance criteria (or succeeds).</p>
  <p>The PRIM algorithm is provided each scenario's parameter values and classification (interesting/uninteresting).</p>
  <p>The algorithm identifies boxes in parameter space. There is a trade-off between coverage, density, and interpretability. High coverage means that the box includes a higher proportion of the ‘interesting’ model scenarios available through parameter space. High density means that the box includes a higher proportion of ‘interesting’ cases relative to non-interesting ones. Some bounds may be more interpretable than others </p>

>   <p>The selection of boxes is usually interactive to address this trade-off, using graphs or constraints on the algorithm to find a box of interest.</p>

Reviewer: It seems a lot has been deleted from this that I would consider useful.

>   <li> a primary 'peeling' stage in which PRIM algortithm explore the input space where the mean of the output is significantly optimised relative to the mean value over the entire dataset. Boxes of increasingly smaller and denser are generated and successively removed from the input space forming a 'peeling trajectory'.[5]</li>

Reviewer: ```suggestion
  <li> a primary 'peeling' stage in which PRIM algortithm explores the input space where the mean of the output is significantly different to the mean value over the entire dataset. Boxes of increasingly smaller and denser are generated and successively removed from the input space forming a 'peeling trajectory' [5].</li>
```
Word missing in "boxes of increasingly small and denser are"

>   <li> a 'pasting' stage could be used to improve the definition of candidate boxes boundaries that best balance objectives of high density and high coverage.[5]</li>

Reviewer: I assume pasting means adding candidate boxes manually, but this would need to  be said explicitly

>   <li> a 'covering' stage in which data points within the identified boxes are removed from the dataset.[5]</li></ol></p>

Reviewer: Not clear why data points would need to be removed

>   <p> Scenarios can be randomly or systematically sampled but need to cover the parameter space well to obtain sensible parameter bounds[1]</p>

Reviewer: Doesn't this need to be said first in order to define what data points are?

    <li> Applying rule extraction to PRIM is used to train a metamodel to identify an hyperbox with high coverage and high density and reduce simulations' run times by 75%.[7] </li>
    <li> Consideration of more than two classes in scenario discovery' s identification of boxes and of integers or categories based on modeified objective functions [9]</li>
tools:
  - slug: sdtoolkit
    name: sdtoolkit
    description: an R package implementing PRIM specifically for scenario discovery
  - slug: emaworkbench
    name: Exploratory Modelling and Analysis (EMA) Workbench
    description: a Python package that includes a scenario discovery oriented implementation of PRIM.
  - slug: primpackage
    name: prim package
    description: non-interactive package written in R programming language [2]
  - slug: openMORDM
    name: openMORDM
    description: Software developed as an R package for generating or loading datasets [3]
also_see:
  - permalink: /tools/CART/
    name: CART
    description: A common alternative algorithm for scenario discovery
further_reading:
  - >

>     [1] Bryant BP, Lempert RJ (2010) <a href="https://dx.doi.org/10.1016/j.techfore.2009.08.002">Thinking inside the box: A participatory, computer-assisted approach to scenario discovery.</a> Technological Forecasting and Social Change, 77(1), 34–49. doi:10.1016/j.techfore.2009.08.002

Reviewer: ```suggestion
references :
  - >
    [1] Bryant BP, Lempert RJ (2010) <a href="https://dx.doi.org/10.1016/j.techfore.2009.08.002">Thinking inside the box: A participatory, computer-assisted approach to scenario discovery.</a> Technological Forecasting and Social Change, 77(1), 34–49. doi:10.1016/j.techfore.2009.08.002
```

  - >
    [13] Groves, D. G., & Bloom, E. (2013). Robust Water-Management Strategies for the California Water Plan Update 2013. RAND Corporation, Santa Monica, CA.
examples:
  - Boxes are shown in a <a href="../../examples/guillaume-2015/">simple flood demonstration problem</a> answering the question "Will regular flooding of ecological assets occur?"
  - Examples are provided using EMA Workbench in its <a href="https://emaworkbench.readthedocs.io/en/latest/indepth_tutorial/open-exploration.html#Scenario-Discovery">documentation</a> and on the <a href="https://waterprogramming.wordpress.com/2015/08/05/scenario-discovery-in-python/">waterprogramming blog</a> (with comparison with CART)
  - Application to the "lake problem" to identify robust water pollution management policies [8]
  - Identification of uncertain factors influential on cooperative regional water planning under deep uncertainty [10]
  - Application for scenario discovery and sensitivity analysis for designing robust operation rules for multi-purpose reservoirs [11]
  - Identification of robust strategies for urban water management [12]

>   - Evaluation of robust water management strategoes for the California Water Plan [13]

Reviewer: Not entirely clear how PRIM is used in each case - would be good to indicate whether boxes are parameters defining policies or external drivers, and perhaps even which variables are used.

    <li> Applying rule extraction to PRIM is used to train a metamodel to identify an hyperbox with high coverage and high density and reduce simulations' run times by 75%.[7] </li>
    <li> Consideration of more than two classes in scenario discovery' s identification of boxes and of integers or categories based on modeified objective functions [9]</li>
tools:
  - slug: sdtoolkit
    name: sdtoolkit
    description: an R package implementing PRIM specifically for scenario discovery
  - slug: emaworkbench
    name: Exploratory Modelling and Analysis (EMA) Workbench
    description: a Python package that includes a scenario discovery oriented implementation of PRIM.
  - slug: primpackage
    name: prim package
    description: non-interactive package written in R programming language [2]
  - slug: openMORDM
    name: openMORDM
    description: Software developed as an R package for generating or loading datasets [3]
also_see:
  - permalink: /tools/CART/
    name: CART
    description: A common alternative algorithm for scenario discovery
further_reading:
  - >

> references :

Reviewer: ```suggestion
    Bryant BP, Lempert RJ (2010) <a href="https://dx.doi.org/10.1016/j.techfore.2009.08.002">Thinking inside the box: A participatory, computer-assisted approach to scenario discovery.</a> Technological Forecasting and Social Change, 77(1), 34–49. doi:10.1016/j.techfore.2009.08.002
references :
```
